New Post about APKBUILD.vim (#13)
Cogitri/cogitri.github.io: content/posts/08-meet-APKBUILD.vim.md

---
title: "APKBUILD.vim: filetype, plugins, syntax and plugins for APKBUILDs on vim"
date: 2020-10-30T17:39:23-03:00
draft: false
---

> Article by editor Leo <thinkabit.ukim@gmail.com>

The Linux distro I contribute to, [Alpine Linux][1], uses build recipes called

> APKBUILDs, they are shell scripts but with context-aware variables that can

Reviewer: ```suggestion
APKBUILDs. They are shell scripts but with context-aware variables that can
```
Otherwise the sentence get really long and hard to follow

> APKBUILDs, which are shell scripts but with context-aware variables that can

Reviewer: ```suggestion
APKBUILDs. They are shell scripts but with context-aware variables that can
```

> affect the final result of package and they aresubject to policies that only

Reviewer: ```suggestion
affect the final result of package. APKBUILDs are subject to policies that only
```

> affect the final result of package and subject to policies that only apply to

Reviewer: ```suggestion
affect the final result of package. They are subject to policies that only apply to
```

apply to [Alpine Linux][1] and its packaging system `abuild`.

> [Vim][3] is a text editor used by many of our developers and users use it
> to edit APKBUILDs. But while vim has a filetype for shellscripts, it is

Reviewer: ```suggestion
[Vim][3] is a text editor used by many of our developers and users for
editing APKBUILDs. But while vim has a filetype for shellscripts, it is
```

> [Vim][3] is a widely used text editor, many of our developers and users use it
> to edit APKBUILDs daily, but while vim has a filetype for shellscripts, it is

Reviewer: ```suggestion
[Vim][3] is a widely used text editor, used by many of our developers and users
to edit APKBUILDs. But while vim has a filetype for shellscripts, it is
```

insufficient to provide syntax, plugins, indentation and other nice to have
features that are specific to the context of an APKBUILD.

In light of that I took upon myself to write a complete implementation of

> a vim `filetype`, specific for APKBUILDs.

Reviewer: ```suggestion
a vim `filetype` for APKBUILDs.
```

Meet [APKBUILD.vim][2].

<!--more-->

## The Whole Deal

[APKBUILD.vim][2] is a plugin for Vim and [Neovim][4], that provides:

1. Syntax Highlighting:  
    1.1. Different colors for variables that affect `abuild` behavior.  

>     1.2. Error highlight for invalid values inside `abuild` variables, like non-existant arches in `arch`.

Reviewer: ```suggestion
    1.2. Error highlighting for invalid values inside `abuild` variables, like non-existant arches in `arch`.
```

Reviewer: Also: https://www.wordnik.com/words/nonexistent

>     1.4. Highlights `secfixes:` blocks, version identifiers and recognized security identifiers (CVE, XSA, etc).  

Reviewer: ```suggestion
    1.4. Highlighting of `secfixes:` blocks, version identifiers and recognized security identifiers (CVE, XSA, etc).  
```

>     1.5. Built on top of `sh.vim` so all syntax highlight from `filetype=sh` is there underneath.

Reviewer: ```suggestion
    1.5. It is built on top of `sh.vim`. That means all the syntax highlighting of `filetype=sh` is still in place.
```

2. Configuration:  
    2.1. Sets correct `textwidth` of 80 columns.  

>     2.2: Sets `noexpandtab` as APKBUILD uses tabs for indentation.  

Reviewer: ```suggestion
    2.2: Sets `noexpandtab` as APKBUILD use tabs for indentation.  
```

    2.3: Sets `softtabstop` and `shiftwidth` to 4.
3. [ALE][5] Integration:  
    3.1. Provides a script to run [shellcheck][6] with APKBUILD aware filtering and register it into ALE.  

>     3.2. Add linters for `apkbuild-lint` and `secfixes-check` from [atools][7] and a handler for them.

Reviewer: ```suggestion
    3.2. Adds support for the linters `apkbuild-lint` and `secfixes-check` from [atools][7] and registers a handler for them.
```

## Pictures

Here are some side by side comparisons of APKBUILDs with `filetype=sh` and `filetype=APKBUILD`:

> Main block of variables used by `abuild`, notice that the variables that are from `abuild`
> have a different color than the ones normally used by shell.

![Comparison of the main block](08-comparison-main-block.png)

> Comparison of secfixes, notice how the `secfixes:` declaration, the versions and the security
> identifiers are all highlighted.

![Comparison of secfixes block](08-comparison-secfixes.png)

> > General comparison of Error highlighting.

Reviewer: ```suggestion
> General comparison of error highlighting.
```

Author: That is correct, Error is a syntax type in Vim

Author: `hi def link InvalidString Error` will make it bright red

Reviewer: Then it should probably be `Error`

Author: done
